File: scripts/updatePublications.pl
#!/usr/bin/env perl
use strict;
use warnings;
use XML::Simple;
use WWW::Curl::Easy;
use JSON::PP qw(encode_json decode_json);
use Data::Dumper;

# Update the `publications.xml` database, adding in any new data (updated NASA ADS bibcodes, DOIs, etc.), and ensurnig that tags
# in the Galacticus repo are also updated.
# Andrew Benson (30-May-2025)

# Read arguments.
die("Usage: updatePublications.pl <apiToken> <slackURL> <repoPath>")
    unless ( scalar(@ARGV) == 3 );
my $apiToken = $ARGV[0];
my $slackURL = $ARGV[1];
my $repoPath = $ARGV[2];

# Extract all tags from the repo.
my %tags;
open(my $git,"git --git-dir ".$repoPath."/.git tag -l |");
while ( my $tag = <$git> ) {
    chomp($tag);
    $tags{$tag} = 0;
}
close($git);

# Parse the database.
my $xml          = new XML::Simple();
my $publications = $xml->XMLin("publications.xml");

# Extract/construct current bibcodes.
print "Searching for bibcodes in database...\n";
my %bibCodes;
foreach my $publication ( @{$publications->{'publication'}} ) {
    my $bibCode;
    if      ( exists($publication->{'bibCode'}) ) {
	# Use the NASA ADS bibcode directly.
	$bibCode = $publication->{'bibCode'};
    } elsif ( exists($publication->{'arXiv'  }) ) {
	# Construct a bibcode from the arXiv number.
	(my $arXivReduced = $publication->{'arXiv'}) =~ s/\.//;
	$bibCode = $publication->{'year'}."arXiv".$arXivReduced.substr($publication->{'author'},1,1);
    }
    # Store a reference to this publication in our bibcodes dictionary.
    if ( defined($bibCode) ) {
	$bibCodes{$bibCode}->{'entry'} = $publication;
	print "   found bibcode: ".$bibCode."\n";
    } else {
	print Dumper($publication);
	die("no bibcode available for this entry");
    }
}
print("...done\n");

# Pull records from NASA ADS.
## Construct a curl object.
my $curl = WWW::Curl::Easy->new();
$curl->setopt(CURLOPT_HEADER,1);
$curl->setopt(CURLOPT_HTTPHEADER, ['Authorization: Bearer '.$apiToken,"Content-Type: big-query/csv"]);
my $countRecords = scalar(keys(%bibCodes));
$curl->setopt(CURLOPT_URL, 'https://api.adsabs.harvard.edu/v1/search/bigquery?q=*:*&rows='.$countRecords.'&fl=bibcode,alternate_bibcode,doi,title,author,year,pub,volume,page');
my $response_body;
$curl->setopt(CURLOPT_WRITEDATA,\$response_body);
$curl->setopt(CURLOPT_POST, 1);
$curl->setopt(CURLOPT_POSTFIELDS, "bibcode\n".join("\n",sort(keys(%bibCodes))));
my $retcode = $curl->perform;
my $records;
if ($retcode == 0) {
    my $response_code = $curl->getinfo(CURLINFO_HTTP_CODE);
    if ( $response_code == 200 ) {
	# Extract the JSON.
	my $json;
	my $startFound = 0;
	open(my $response,"<",\$response_body);
	while ( my $line = <$response> ) {
	    $startFound = 1
		if ( $line =~ m/^\{/ );
	    $json .= $line
		if ( $startFound );
	}
	close($response);
	$records = decode_json($json);
    } else {
	die("Failed to retrieve record identifiers: ".$response_code.$response_body."\n");
    }
} else {
    die("Failed to retrieve record identifiers: ".$retcode." ".$curl->strerror($retcode)." ".$curl->errbuf."\n");
}

# Match records with our entries.
print "Matching NASA ADS bibcodes to entries...\n";
foreach my $record ( @{$records->{'response'}->{'docs'}} ) {
    my $matchedBibCode;
    if ( exists($bibCodes{$record->{'bibcode'}}) ) {
	$matchedBibCode = $record->{'bibcode'};
    } elsif ( exists($record->{'alternate_bibcode'}) ) {
	foreach my $altBibCode ( @{$record->{'alternate_bibcode'}} ) {
	    if ( exists($bibCodes{$altBibCode}) ) {
		$matchedBibCode = $altBibCode;
	    }
	}
    }
    if ( defined($matchedBibCode) ) {
	$bibCodes{$matchedBibCode}->{'record'} = $record;
	print "   matched record for bibcode: ".$matchedBibCode."\n";
    } else {
	die("unable to match record (bibcode: ".$record->{'bibcode'}.") to entry");
    }
}
print "...done\n";

# Process each entry and apply updates.
print "Checking for updates...\n";
my @newTags;
foreach my $bibCode ( keys(%bibCodes) ) {
    # Replace the bibcode with the canonical bibcode.
    if ( $bibCodes{$bibCode}->{'record'}->{'bibcode'} ne $bibCodes{$bibCode}->{'entry'}->{'bibCode'} ) {
	print "   update bibcode to canonical value: ".$bibCodes{$bibCode}->{'entry'}->{'bibCode'}." --> ".$bibCodes{$bibCode}->{'record'}->{'bibcode'}."\n";
	$bibCodes{$bibCode}->{'entry'}->{'bibCode'} = $bibCodes{$bibCode}->{'record'}->{'bibcode'};
    }
    # Add DOI if necessary.
    foreach my $doi ( @{$bibCodes{$bibCode}->{'record'}->{'doi'}} ) {
	next
	    if ( $doi =~ m/arXiv/ );
	if ( $doi ne $bibCodes{$bibCode}->{'entry'}->{'doi'} ) {
	    $bibCodes{$bibCode}->{'entry'}->{'doi'} = $doi;
	    print "   update doi for bibcode ".$bibCodes{$bibCode}->{'entry'}->{'bibCode'}." to: ".$doi."\n";
	}
	last;
    }
    # Add journal URL if necessary.
    if ( exists($bibCodes{$bibCode}->{'entry'}->{'doi'}) ) {
	my $curl = WWW::Curl::Easy->new();
	$curl->setopt(CURLOPT_URL, 'https://doi.org/'.$bibCodes{$bibCode}->{'entry'}->{'doi'});
	my $response_body;
	$curl->setopt(CURLOPT_WRITEDATA,\$response_body);
	my $retcode = $curl->perform;
	if ($retcode == 0) {
	    my $journalURL;
	    open(my $response,"<",\$response_body);
	    while ( my $line = <$response> ) {
		if ( $line =~ m/href="(.+)"/ ) {
		    my $url = $1;
		    $journalURL = $url
			unless ( $url =~ m/arxiv/i );
		}
	    }
	    close($response);
	    if ( defined($journalURL) ) {
		if ( $journalURL ne $bibCodes{$bibCode}->{'entry'}->{'journalURL'} ) {
		    $bibCodes{$bibCode}->{'entry'}->{'journalURL'} = $journalURL;
		    print "   update journal URL for bibcode ".$bibCodes{$bibCode}->{'entry'}->{'bibCode'}." to: ".$journalURL."\n";
		}
	    }
	} else {
	    die("Failed to resolve DOI for bibcode ".$bibCodes{$bibCode}->{'entry'}->{'bibCode'}.": ".$retcode." ".$curl->strerror($retcode)." ".$curl->errbuf."\n");
	}
    }
    # Update repo tags.
    if ( exists($bibCodes{$bibCode}->{'entry'}->{'commit'}) ) {
	my $message = "\"Publication: ".$bibCodes{$bibCode}->{'entry'}->{'title'}." by ".$bibCodes{$bibCode}->{'entry'}->{'author'}." (".$bibCodes{$bibCode}->{'entry'}->{'year'}.")\"";
	if ( exists($bibCodes{$bibCode}->{'entry'}->{'arXiv'}) ) {
	    my $tag = "publication/arXiv/".$bibCodes{$bibCode}->{'entry'}->{'arXiv'};
	    if ( exists($tags{$tag}) ) {
		++$tags{$tag};
	    } else {
		push(@newTags,"git tag -a ".$tag." ".$bibCodes{$bibCode}->{'entry'}->{'commit'}." -m ".$message);
	    }
	}
	if ( exists($bibCodes{$bibCode}->{'entry'}->{'bibCode'}) ) {
	    (my $bibcodeGitified = $bibCodes{$bibCode}->{'entry'}->{'bibCode'}) =~ s/\./_/g;
	    my $tag = "publication/bibcode/".$bibcodeGitified;
	    if ( exists($tags{$tag}) ) {
		++$tags{$tag};
	    } else {
		push(@newTags,"git tag -a ".$tag." ".$bibCodes{$bibCode}->{'entry'}->{'commit'}." -m ".$message);
	    }
	}
	if ( exists($bibCodes{$bibCode}->{'entry'}->{'doi'}) ) {
	    my $tag = "publication/doi/".$bibCodes{$bibCode}->{'entry'}->{'doi'};
	    if ( exists($tags{$tag}) ) {
		++$tags{$tag};
	    } else {
		push(@newTags,"git tag -a ".$tag." ".$bibCodes{$bibCode}->{'entry'}->{'commit'}." -m ".$message);
	    }
	}
    }
}
# Add deletes for any obsoleted tags.
foreach my $tag ( keys(%tags) ) {
    push(@newTags,"git tag -d ".$tag)
	if ( $tags{$tag} == 0 && $tag =~ m/^publication\// );
}
print "...done\n";

# Write updated database file.
open(my $database,">","publications.xml");
print $database $xml->XMLout($publications,RootName => "publications");
close($database);

# Send a Slack notification for any new tags needed.
if ( scalar(@newTags) > 0 ) {
    print "New tags to be created:\n\t".join("\n\t",@newTags)."\n";
    my $data =
    {
	tags => join("\n",@newTags)."\n"
    };
    my $json = encode_json($data);
    system("curl -X POST -H 'Content-type: application/json' --data '".$json."' ".$slackURL);
} else {
    print "No new tags to be created\n";
}

exit;

File: scripts/updatePublications.py
#!/usr/bin/env python3
"""Update the `publications.xml` database, adding in any new data
(updated NASA ADS bibcodes, DOIs, etc.), and ensuring that tags in
the Galacticus repo are also updated.

Andrew Benson (30-May-2025; ported to Python 09-May-2026)
"""

import json
import subprocess
import sys
import urllib.error
import urllib.request
import xml.etree.ElementTree as ET


def main():
    if len(sys.argv) != 4:
        sys.exit("Usage: updatePublications.py <apiToken> <slackURL> <repoPath>")
    api_token, slack_url, repo_path = sys.argv[1:4]

    # Extract all tags from the repo.
    print("Looking for existing tags...")
    tags = {}
    git_output = subprocess.check_output(
        ["git", "--git-dir", f"{repo_path}/.git", "tag", "-l"],
        text=True,
    )
    for tag in git_output.splitlines():
        if not tag:
            continue
        tags[tag] = 0
        if tag.startswith("publication/"):
            print(f" found: {tag}")
    print("...done")

    # Parse the database.
    tree = ET.parse("publications.xml")
    publications = list(tree.getroot().findall("publication"))

    # Extract/construct current bibcodes.
    print("Searching for bibcodes in database...")
    bibcodes = {}
    for publication in publications:
        attrib = publication.attrib
        if "bibCode" in attrib:
            bibcode = attrib["bibCode"]
        elif "arXiv" in attrib:
            arxiv_reduced = attrib["arXiv"].replace(".", "")
            bibcode = f"{attrib['year']}arXiv{arxiv_reduced}{attrib['author'][0]}"
        else:
            print(ET.tostring(publication, encoding="unicode"))
            sys.exit("no bibcode available for this entry")
        bibcodes[bibcode] = {"entry": publication}
        print(f"   found bibcode: {bibcode}")
    print("...done")

    # Pull records from NASA ADS.
    body = ("bibcode\n" + "\n".join(sorted(bibcodes.keys()))).encode("utf-8")
    url = (
        "https://api.adsabs.harvard.edu/v1/search/bigquery"
        f"?q=*:*&rows={len(bibcodes)}"
        "&fl=bibcode,alternate_bibcode,doi,title,author,year,pub,volume,page"
    )
    request = urllib.request.Request(
        url,
        data=body,
        headers={
            "Authorization": f"Bearer {api_token}",
            "Content-Type": "big-query/csv",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(request) as response:
            response_body = response.read().decode("utf-8")
    except urllib.error.HTTPError as error:
        detail = error.read().decode("utf-8", errors="replace")
        sys.exit(f"Failed to retrieve record identifiers: {error.code}{detail}")
    except urllib.error.URLError as error:
        sys.exit(f"Failed to retrieve record identifiers: {error.reason}")
    records = json.loads(response_body)

    # Match records with our entries.
    print("Matching NASA ADS bibcodes to entries...")
    for record in records["response"]["docs"]:
        matched = None
        if record["bibcode"] in bibcodes:
            matched = record["bibcode"]
        else:
            for alt in record.get("alternate_bibcode", []):
                if alt in bibcodes:
                    matched = alt
        if matched is None:
            sys.exit(f"unable to match record (bibcode: {record['bibcode']}) to entry")
        bibcodes[matched]["record"] = record
        print(f"   matched record for bibcode: {matched}")
    print("...done")

    # Process each entry and apply updates.
    print("Checking for updates...")
    new_tags = []
    for bibcode, info in bibcodes.items():
        if "record" not in info:
            print(f"   FAIL: no ADS record found for bibcode '{bibcode}'")
            sys.exit(1)
        entry = info["entry"]
        record = info["record"]

        # Replace the bibcode with the canonical bibcode.
        if record["bibcode"] != entry.get("bibCode"):
            print(
                f"   update bibcode to canonical value: {entry.get('bibCode')} "
                f"--> {record['bibcode']}"
            )
            entry.set("bibCode", record["bibcode"])

        # Add DOI if necessary.
        for doi in record.get("doi", []):
            if "arXiv" in doi:
                continue
            if doi != entry.get("doi"):
                entry.set("doi", doi)
                print(f"   update doi for bibcode {entry.get('bibCode')} to: {doi}")
            break

        # Add journal URL if necessary. Resolve the DOI via doi.org's
        # handle API, which returns the publisher URL as structured JSON.
        if "doi" in entry.attrib:
            try:
                with urllib.request.urlopen(
                    f"https://doi.org/api/handles/{entry.get('doi')}"
                ) as response:
                    handle = json.loads(response.read())
            except urllib.error.URLError as error:
                sys.exit(
                    f"Failed to resolve DOI for bibcode {entry.get('bibCode')}: "
                    f"{error.reason}"
                )
            journal_url = next(
                (
                    value["data"]["value"]
                    for value in handle.get("values", [])
                    if value.get("type") == "URL"
                    and "arxiv" not in value["data"]["value"].lower()
                ),
                None,
            )
            if journal_url is not None and journal_url != entry.get("journalURL"):
                entry.set("journalURL", journal_url)
                print(
                    f"   update journal URL for bibcode {entry.get('bibCode')} "
                    f"to: {journal_url}"
                )

        # Update repo tags.
        if "commit" in entry.attrib:
            message = (
                f'"Publication: {entry.get("title")} by {entry.get("author")} '
                f'({entry.get("year")})"'
            )
            commit = entry.get("commit")
            for kind, value in (
                ("arXiv", entry.get("arXiv")),
                ("bibcode", (entry.get("bibCode") or "").replace(".", "_") or None),
                ("doi", entry.get("doi")),
            ):
                if not value:
                    continue
                tag = f"publication/{kind}/{value}"
                if tag in tags:
                    tags[tag] += 1
                else:
                    new_tags.append(f"git tag -a {tag} {commit} -m {message}")

    # Add deletes for any obsoleted tags.
    for tag, count in tags.items():
        if count == 0 and tag.startswith("publication/"):
            new_tags.append(f"git tag -d {tag}")
    print("...done")

    # Write updated database file.
    write_publications("publications.xml", publications)

    # Send a Slack notification for any new tags needed.
    if new_tags:
        print("New tags to be created:\n\t" + "\n\t".join(new_tags))
        payload = json.dumps({"tags": "\n".join(new_tags) + "\n"})
        result = subprocess.run(
            [
                "curl", "-X", "POST",
                "-H", "Content-type: application/json",
                "--data", payload,
                slack_url,
            ]
        )
        if result.returncode != 0:
            print(
                f"Warning: curl exited with status {result.returncode} "
                f"when sending Slack notification",
                file=sys.stderr,
            )
    else:
        print("No new tags to be created")


def write_publications(path, publications):
    """Write publications XML in a format matching XML::Simple's XMLout."""
    lines = ["<publications>"]
    for publication in publications:
        attrs = " ".join(
            f'{name}="{_xml_escape(value)}"'
            for name, value in sorted(publication.attrib.items())
        )
        lines.append(f"  <publication {attrs} />")
    lines.append("</publications>")
    with open(path, "w", encoding="utf-8") as handle:
        handle.write("\n".join(lines) + "\n")


def _xml_escape(value):
    return (
        value.replace("&", "&amp;")
             .replace("<", "&lt;")
             .replace(">", "&gt;")
             .replace('"', "&quot;")
    )


if __name__ == "__main__":
    main()
